Fix transaction propagation

SWvheerden opened the pull request:
Set bootstrap default false

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This fixes the transaction propagation and the bootstrap flag.
Fixes transaction builder in cucumber. 
Adds in a cucumber test to test transaction propagation.
Adds cucumber test for double spends. 
Adds cucumber test for mempool clearing
Adds in a GRPC call to search the mempool for a transaction excess signature. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Transactions are sometimes sent and they dont propagate through he network. The Bootstrap flag is set to false everytime the node enters sync. When the bootstrap flag is set, the node will ignore all blocks and transactions sent to it. It should only ignore these during initial sync and not every sync. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Integration tests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

delta1 commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -264,3 +264,19 @@ message GetMempoolTransactionsRequest {
+}
+
+enum TransactionStateResult {
+    UNKOWN = 0;
```
spelling: `UNKNOWN`

philipr-za commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -264,3 +264,19 @@ message GetMempoolTransactionsRequest {
+    TransactionStateResult result =1;
+}
+
+enum TransactionStateResult {
```
I would call this TransactionLocation, Reorg is not a state?

philipr-za commented on base_layer/core/src/base_node/state_machine_service/states/events_and_states.rs:
```diff
@@ -216,7 +216,7 @@ pub struct StatusInfo {
     pub fn new() -> Self {
         Self {
-            bootstrapped: false,
+            bootstrapped: true,
```
This doesn't make sense. The whole point of Bootstrapped is to indicate that the node has not synced even once since startup. That is why when you first start the node it is in the StartUp state and Bootstrapped = false. It should only flip to True once it has synced at least once.

philipr-za replied:
This is the code that should flip it to true inside the Listening State:
```Rust
if sync_mode.is_lagging() {
                        return StateEvent::FallenBehind(sync_mode);
                    } else {
                        if !shared.bootstrapped_sync && sync_mode == SyncStatus::UpToDate {
                            shared.bootstrapped_sync = true;
                            debug!(
                                target: LOG_TARGET,
                                "Initial sync achieved, bootstrap done: {}", sync_mode
                            );
                        }
                        self.is_synced = true;
                        shared.set_state_info(StateInfo::Listening(ListeningInfo::new(true)));
                    }
```

SWvheerden replied:
I Think this might have slipped through. I have been changing values left and right to get it to work correctly

philipr-za commented on base_layer/core/src/base_node/state_machine_service/states/block_sync.rs:
```diff
@@ -74,7 +74,7 @@ impl BlockSync {
 
             let _ = status_event_sender.broadcast(StatusInfo {
-                bootstrapped: false,
+                bootstrapped: true,
```
This should not be false (same for in the HeaderSync state) but just setting this to true nullifies the whole purpose of the Bootstrapped flag. The current value should be drawn from the Shared state value `shared.bootstrapped_sync`

SWvheerden replied:
I think this comes down to, should header sync change this value?

philipr-za replied:
nope, only the first time you get to Listening should change it

philipr-za reviewed (commented):
This PR breaks the Bootstrapped flag

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -80,6 +80,13 @@ message NewBlockTemplateResponse{
+/// return type of GetNewBlockTemplate
+message NewBlockTemplateRequest{
+    PowAlgo algo = 1;
+    //This field should be moved to optional once optional keyword is standard
```
Comment is not needed since all fields are optional in protobuf

SWvheerden replied:
Fields for messages are optional, but not primitives:
This was released 5 days ago:
https://github.com/protocolbuffers/protobuf/releases/tag/v3.15.0

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -80,6 +80,13 @@ message NewBlockTemplateResponse{
     MinerData miner_data = 4;
 }
 
+/// return type of GetNewBlockTemplate
```
Comment looks like a bad copy pasta

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -264,3 +271,19 @@ message GetMempoolTransactionsRequest {
 }
+
+message TransactionStateRequest {
+    Transaction transaction  = 1;
```
Seems heavy to send a whole transaction in. Perhaps just the excess or a hash of a kernel?

stringhandler commented on applications/tari_app_grpc/proto/base_node.proto:
```diff
@@ -264,3 +271,19 @@ message GetMempoolTransactionsRequest {
+
+enum TransactionLocation {
+    UNKNOWN = 0;
+    UNCONFIRMED = 1;
```
Unconfirmed should return 'Mempool' and REORG should probably map to 'NotStored' or 'Mined'. The caller does not care about this internal details. @philipr-za what do you think?

philipr-za replied:
This is the route taken in the Base Node RPC service exposed to the wallet because we want to keep the idea of the Reorg pool as an internal idea. So I would think that this should also be the case in the gRPC which is an external interface. I understand that in the test you want to check how the system reacts to Double Spends vs Orphans but that can be done in an internal Rust test while the Cucumber test should be an Integration test that tests the system as if you are interacting with it from the outside as it will be in the normal operation. 

In the Cucumber test you can create a double spend and see that it is rejected correctly.

stringhandler commented on applications/tari_base_node/src/grpc/base_node_grpc_server.rs:
```diff
@@ -529,6 +535,54 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
+                result: tari_rpc::TransactionLocation::Reorg.into(),
+            },
+            TxStorageResponse::NotStoredAlreadySpent => tari_rpc::TransactionStateResponse {
+                result: tari_rpc::TransactionLocation::Mined.into(),
```
I don't think this mapping is correct

SWvheerden replied:
Its what the mempool sends back if a transaction is already spent. If its spent, it has to be mined.

stringhandler commented on applications/tari_base_node/src/grpc/base_node_grpc_server.rs:
```diff
@@ -529,6 +535,54 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
+                result: tari_rpc::TransactionLocation::Unconfirmed.into(),
+            },
+            TxStorageResponse::ReorgPool => tari_rpc::TransactionStateResponse {
+                result: tari_rpc::TransactionLocation::Reorg.into(),
```
I think this is more likely that the tx is Mined at this point. I'm sure there is logic similar to this in the base node already, perhaps on the RPC side

SWvheerden replied:
The wallets have proper mapping for this. But the BaseNode's not. All the test just assume if the mempool says accepted, its in the mempool and or mined. .

stringhandler commented on base_layer/core/src/base_node/comms_interface/comms_request.rs:
```diff
@@ -53,7 +53,7 @@ pub enum NodeCommsRequest {
     FetchBlocksWithStxos(Vec<Commitment>),
     FetchBlocksWithUtxos(Vec<Commitment>),
-    GetNewBlockTemplate(PowAlgorithm),
+    GetNewBlockTemplate(PowAlgorithm, u64),
```
This should be changed to a struct with named params

stringhandler commented on base_layer/core/src/mempool/sync_protocol/mod.rs:
```diff
@@ -132,11 +136,28 @@ where TSubstream: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static
+                        target: LOG_TARGET,
+                        "Mempool sync still on hold, waiting for bootstrap to finish",
+                    );
+                    tokio::time::delay_for(Duration::from_secs(60)).await;
```
This would be more efficient as a `select!` macro with a timeout

stringhandler commented on integration_tests/features/Mempool.feature:
```diff
@@ -0,0 +1,83 @@
+    Given I have 10 seed nodes
+    And I have a base node SENDER connected to all seed nodes
+    And I have 10 base nodes connected to all seed nodes
+    When I wait 100 seconds
```
These waits need justifications in comments so that they can be improved upon in future. Describe why the wait is 100 seconds so that a future dev can know where to start optimizing

stringhandler reviewed (commented):
I like the tests and the concept, but the transaction state should ask both the mempool and the chain for where a transaction is. There are lots of other style nits and a mapping error. 

I'm sure this exact task of asking where a transaction kernel is has been done in the base node RPC, but I might be wrong. The logic should not differ between these implementations

philipr-za commented on integration_tests/features/Mempool.feature:
```diff
@@ -0,0 +1,83 @@
+    When I submit transaction TX1 to SENDER
+    Then SENDER has TX1 in its mempool inside of UNCONFIRMED pool
+    When I wait 100 seconds
+    Then TX1 is in the UNCONFIRMED pool of all nodes
```
It is new stuff Hansie and I have started to use in the Cucumber steps but look at the `I wait` type steps in https://github.com/tari-project/tari/pull/2681 and https://github.com/tari-project/tari/pull/2686 to see a better way to do this with the time constraint so you don't have to wait a 100 second each time but wait until the condition has been met OR the max time has passed.

SWvheerden committed:
add GPRC call to search mempool

Set bootstrap default false
add mempool sync cucumber test
add cucumber test for mempool clearing by fee
Add double spend cucumber test
